### ISSUE TITLE
Large memory access causes execution to fail.

### ISSUE THREAD

**Issue by Subway2023** [paritytech/revive#267]
Resolc version: 0.1.0-dev.12+commit.76f4cf7.llvm-18.1.8
Description: The instruction **let memPtr := add(mload(0x40), 65536)** causes execution to fail, but solc+evm does not report an error. The [documentation ](https://contracts.polkadot.io/revive_compiler/differences_yul_translation/#mload-mstore-msize-mcopy-memory-related-functions)does not clearly explain this difference, and it is unclear where the boundary for memory access lies.

```solidity
contract TestCompiler {
    function test() public pure returns (bytes1) {
        bytes1 result;
        assembly {
            let memPtr := add(mload(0x40), 65536)
            mstore8(memPtr, 0xab)
            result := mload(sub(memPtr, 31))
        }
        return result;
    }
}
```
### execute
I have set up an execution [environment ](https://github.com/Subway2023/revive)that can directly execute the bytecode generated by revive compilation.
```
revive/target/debug/revive-runner 50564d00000906000000000000010700c11000c00040047106000000000e0000001c0000002a000000350000004700000063616c6c5f646174615f636f707963616c6c5f646174615f6c6f616463616c6c5f646174615f73697a657365616c5f72657475726e7365745f696d6d757461626c655f6461746176616c75655f7472616e73666572726564051102849f0463616c6c84a6066465706c6f79068569100284af5a005f00780009011e0143014f016f014b025c029502c60321048e04a604ae048b791052093b3309010001ae9833c8780883881f8488e0553801000128390a000002ae8a123d080000023308100002c8870732003308100002c88707320000009511f07b10087b158475010a02013d0704000251050750100209501004f5009511c07b10387b15309515408411e03307403308205010068c46171f800046171e46171d46171c46171b46171a46171946171846171746171646171546171446171346171246171146171046170f46170e46170d46170c46170b46170a4617094617084617074617064617054617044617034617024617014607491118490149110849111083170a0501821710821818821908821ad49808d4a707d4870751071633073308501008fcfe837833070133090a030133078000330850100ae7fe8377330833090a01390808000251080d330710000383770a040133078000330850100cc2fe83783307330933000e0a03019511807b10787b15707b1668951580008411e0641633074033082050101096fe46171f800046171e46171d46171c46171b46171a46171946171846171746171646171546171446171346171246171146171046170f46170e46170d46170c46170b46170a461709461708461707461703330804000246170246170146077b68383908040002461706461705461704531804e00183674033080a010180675c9777201408000000006dfda8f8ab87c3019517e08477e07b67306471491718491710491708490783770a0501826a3082a71082a81882a90882aad49808d4a707d4870752078d01826738817754170384013307800001330801501012bafd4607ab003307610001330820501014a9fd7c787b68207c78017b687c78027b68107c78037b68087c78047b68387c78057b68187c78067b68307c77077b672833074033082050101670fd7c78117c79107c7a127c7b13978808d4980897aa1097bb18d4ba0ad4a8087c79157c7a147c7b167c7c17979908d4a90997bb1097cc18d4cb0bd4b909979920d489027c79017c7a7c7b027c7c03979908d4a90997bb1097cc18d4cb0bd4b9097c7a057c7b047c7c067c780797aa08d4ba0a97cc10978818d4c808d4a808978820d498037c78197c7a187c7b1a7c7c1b978808d4a80897bb1097cc18d4cb0bd4b8087c7a1d7c7b1c7c7c1e7c791f97aa08d4ba0a97cc10979918d4c909d4a909979920d4890a7c78097c79087c7b0a7c7c0b978808d4980897bb1097cc18d4cb0bd4b8087c790d7c7b0c7c7c0e7c770f979908d4b90997cc10977718d4c707d49707977720d487076f786fa76f396f2ad4a909d4870ad49a0a98aa20d49808978820d4a80851081700330733085010183ffc837833070133090a03018268978808826920d49808826910979910826a0897aa18d4a909d49808826918979908826a38d4a909826a3097aa10826b2897bb18d4ba0ad4a909979920d498087b683883777b673033082050101ae4fb46171f46171e46171d46171c46171b46171a46171946171846171746171646171546171446171346171246171146171046170f46170e46170d46170c46170b46170a461709461708461707461706461705461704461703461702461701826838787833082082673050101c77fb837833092033070a03330733001e28a4fb330750101e9ffb33070150102097fb0049489220248414d2496948842449122149922449922449922449925252939424154a8d427548a851a8d2248994849024499224499224499290949024a426014492925493a42491424888905052922449922421499224499224499294244992244992244992244992244992244992244992244952952449b242a9254992244992244952129224499224499224499224494a42a992424200 f8a8fd6d
```
```
assertion `left == right` failed: contract execution result mismatch: Exec { result: ContractResult { gas_consumed: Weight { ref_time: 6218497, proof_size: 1545 }, gas_required: Weight { ref_time: 6218497, proof_size: 1545 }, storage_deposit: StorageDeposit::Charge(0), result: Err(Module(ModuleError { index: 3, error: [10, 0, 0, 0], message: Some("ContractTrapped") })) }, wall_time: 1.130295ms }
  left: true
 right: false
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Comment by xermicus** [comment]
This is expected: Unlike on EVM where memory is capped by gas limits, we emulate teh EVM heap memory with a static buffer. This buffer defaults to 64kb and is configurable with the `--heap-size` argument.
